imports/plugins/core/product/server/no-meteor/mutations/cloneProductVariants.js

```diff
@@ -0,0 +1,124 @@
+import _ from "lodash";
+import SimpleSchema from "simpl-schema";
+import Logger from "@reactioncommerce/logger";
+import Random from "@reactioncommerce/random";
+import ReactionError from "@reactioncommerce/reaction-error";
+import copyMedia from "../utils/copyMedia";
+
+const inputSchema = new SimpleSchema({
+  "shopId": String,
+  "variantIds": Array,
+  "variantIds.$": {
+    type: String
+  }
+});
+
+/**
+ *
+ * @method cloneProductVariants
+ * @summary clones a product variant into a new variant
+ * @description the method copies variants, but will also create and clone
+   * child variants (options)
+ * @param {Object} context -  an object containing the per-request state
+ * @param {Object} input - Input arguments for the bulk operation
+ * @param {String} input.variantIds - an array of decoded variant IDs to clone
+ * @return {Array} list with cloned variant Ids
+ */
+export default async function cloneProductVariants(context, input) {
+  inputSchema.validate(input);
+  const { collections, userHasPermission } = context;
+  const { Products } = collections;
+  const { shopId, variantIds } = input;
+
+  if (!userHasPermission(["createProduct", "product/admin", "product/clone"], shopId)) {
+    throw new ReactionError("access-denied", "Access Denied");
+  }
+
+  // Check to make sure all variants are on the same shop
+  const count = await Products.find({ _id: { $in: variantIds }, type: "variant", shopId }).count();
+  if (count !== variantIds.length) throw new ReactionError("not-found", "One or more variants do not exist");
+
+  const newVariantIds = await Promise.all(variantIds.map(async (variantId) => {
```

Comment:
@aldeed I'm doing a lot of awaiting promises and in here, this works, but not sure if it's the best way to do all of this.